peps/pep-0750.rst

```diff
@@ -1,889 +1,1213 @@
+        return value
 
-Expression evaluation for interpolations is the same as in :pep:`498#expression-evaluation`,
-except that all expressions are always implicitly wrapped with a ``lambda``:
 
-    The expressions that are extracted from the string are evaluated in the context
-    where the tag string appeared. This means the expression has full access to its
-    lexical scope, including local and global variables. Any valid Python expression
-    can be used, including function and method calls.
+    def f(template: Template) -> str:
+        parts = []
+        for arg in template.args:
+            match arg:
+                case str() as s:
+                    parts.append(s)
+                case Interpolation(value, _, conv, format_spec):
+                    value = convert(value, conv)
+                    value = format(value, format_spec)
+                    parts.append(value)
+        return "".join(parts)
 
-However, there's one additional nuance to consider, `function scope
-<https://docs.python.org/3/reference/executionmodel.html#resolution-of-names>`_
-versus `annotation scope
-<https://docs.python.org/3/reference/executionmodel.html#annotation-scopes>`_.
-Consider this somewhat contrived example to configure captions:
 
-.. code-block:: python
+.. note:: Example code
 
-    class CaptionConfig:
-        tag = 'b'
-        figure = f'<{tag}>Figure</{tag}>'
+   See `fstring.py`__ and `test_fstring.py`__.
 
-Let's now attempt to rewrite the above example to use tag strings:
+   __ https://github.com/davepeck/pep750-examples/blob/main/pep/fstring.py
+   __ https://github.com/davepeck/pep750-examples/blob/main/pep/test_fstring.py
 
-.. code-block:: python
 
-    class CaptionConfig:
-        tag = 'b'
-        figure = html'<{tag}>Figure</{tag}>'
+Example: Structured Logging
+---------------------------
+
+Structured logging allows developers to log data in both a human-readable format
+*and* a structured format (like JSON) using only a single logging call. This is
+useful for log aggregation systems that process the structured format while
+still allowing developers to easily read their logs.
+
+We present two different approaches to implementing structured logging with
+template strings.
 
-Unfortunately, this rewrite doesn't work if using the usual lambda wrapping to
-implement interpolations, namely ``lambda: tag``. When the interpolations are
-evaluated by the tag function, it will result in ``NameError: name 'tag' is not
-defined``. The root cause of this name error is that ``lambda: tag`` uses function scope,
-and it's therefore not able to use the class definition where ``tag`` is
-defined.
+Approach 1: Custom Log Messages
+'''''''''''''''''''''''''''''''
 
-Desugaring how the tag string could be evaluated will result in the same
-``NameError`` even using f-strings; the lambda wrapping here also uses function
-scoping:
+The `Python Logging Cookbook <https://docs.python.org/3/howto/logging-cookbook.html>`_
```

Comment:
```suggestion
The :ref:`Python Logging Cookbook <python:logging-cookbook>`
```

Reply:
Thank you for this. Also: I should really have learned this syntax by now! :-)